File: android/src/main/java/com/margelo/nitro/spiralimage/SpiralImage.kt
package com.margelo.nitro.spiralimage

import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import com.margelo.nitro.spiralimage.pipeline.ImagePipeline
import java.io.File

@DoNotStrip
class SpiralImage : HybridSpiralImageSpec() {

    override fun process(
        options: ProcessOptions
    ): Promise<ImageResult> {

        return Promise.async {

            val outputPath =
                options.output?.path
                    ?: throw IllegalArgumentException("output.path is required")

            val cacheDir =
                File(outputPath).parentFile
                    ?: throw IllegalArgumentException("Invalid output.path")

            ImagePipeline.process(
                cacheDir,
                options
            )
        }
    }
}

File: ios/SpiralImage.swift
import Foundation
import NitroModules

final class SpiralImage: HybridSpiralImageSpec {

    func process(
        options: ProcessOptions
    ) throws -> Promise<ImageResult> {

        return Promise.async {
            try ImagePipeline.execute(
                options: ResizeOptions(
                    input: options.path,
                    output: options.output?.path ?? options.path,
                    width: options.resize?.width,
                    height: options.resize?.height,
                    quality: options.output?.quality,
                    format: options.output?.format.map { format in
                        switch format {
                        case .jpeg:
                            return .jpeg
                        case .png:
                            return .png
                        case .webp:
                            return .webp
                        }
                    }
                )
            )
        }
    }
}